core/src/test/java/kafka/server/share/SharePartitionTest.java

```diff
@@ -6190,6 +6198,289 @@ public void testFindLastOffsetAcknowledgedWhenGapAtBeginning() {
         assertEquals(-1, lastOffsetAcknowledged);
     }
 
+    /**
+     * Test the case where the fetch batch has first record offset greater than the record batch start offset.
+     * Such batches can exist for compacted topics.
+     */
+    @Test
+    public void testAcquireAndAcknowledgeWithRecordsAheadOfRecordBatchStartOffset() {
+        SharePartition sharePartition = SharePartitionBuilder.builder()
+            .withState(SharePartitionState.ACTIVE)
+            .build();
+
+        ByteBuffer buffer = ByteBuffer.allocate(4096);
+        // Set the base offset at 5.
+        try (MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, Compression.NONE,
+            TimestampType.CREATE_TIME, 5, 2)) {
+            // Append records from offset 10.
+            memoryRecords(2, 10).records().forEach(builder::append);
+            // Append records from offset 15.
+            memoryRecords(2, 15).records().forEach(builder::append);
+        }
+        buffer.flip();
+        MemoryRecords records = MemoryRecords.readableRecords(buffer);
+        // Complete batch from 5-16 will be acquired, hence 12 records.
+        fetchAcquiredRecords(sharePartition, records, 12);
+        // Partially acknowledge the batch from 5-16.
+        sharePartition.acknowledge(MEMBER_ID, Arrays.asList(
+            new ShareAcknowledgementBatch(5, 9, List.of((byte) 0 /* GAP */)),
+            new ShareAcknowledgementBatch(10, 11, List.of((byte) 1 /* ACCEPT */)),
+            new ShareAcknowledgementBatch(12, 14, List.of((byte) 3 /* GAP */)),
```

Comment:
3 != GAP.

Reply:
I think we probably ought to define constants for these 0,1,2,3 and use them across all components.

Reply:
Corrected, and used constants. I will update other tests with these values while doing a refactor for SharePartitionTest in some future PR, currently corrected the new added tests.